### PR TITLE
cmd/gossamer, consensus/babe, core, state, trie: state genesis initialization, setup babe from core

### LINKED ISSUE #526: state: initialize from genesis
currently, there is no way to initialize the state service with the genesis block. need to add a method for initialization. ideally this will happen when `gossamer init` is done and the genesis data is loaded.

### PR DESCRIPTION
**Pull request by noot**
## Changes
- add function for loading state service from a genesis state. this will be called once, when `gossamer init` is done
- add state service to core, babe so that blocks can be added to it
- setup core so that it spawns new babe sessions
- change babe.invokeBlockAuthoring to actually call buildBlock

## Tests:
```
go test ./...
```

### Issues:
closes #526 
a step towards #565, but doesn't close it yet

### CODE REVIEW

**Comment by ansermino** [comment]
@noot This should be put on ice until we finish #503 and close #465. Services should not import state, they should define their own state interface.

**Comment by noot** [comment]
@ansermino I've updated the PR to have BlockState interfaces in core and babe, no more direct use of state.Service. in the meantime maybe we can discuss how I can help finish the state refactor? since all babe functionality is done

**Comment by noot** [comment]
on ice for now, I'll work on the issues David mentioned first

File: cmd/gossamer/genesis.go [hunk 1]
```diff
@@ -35,32 +38,37 @@ func loadGenesis(ctx *cli.Context) error {
...
-	}
-
-	// create and load storage trie with initial genesis state
-	t := trie.NewEmptyTrie(tdb)
-
-	err = t.Load(gen.GenesisFields().Raw)
-	if err != nil {
-		return fmt.Errorf("cannot load trie with initial state: %s", err)
-	}
+	// setup trie database
```
Comment (ryanchristo) [inline cmd/gossamer/genesis.go]: "set up"

File: consensus/babe/types.go [hunk 2]
```diff
@@ -46,36 +43,11 @@ type AuthorityData struct {
...
-	enc = append(enc, buf...)
-	binary.LittleEndian.PutUint64(buf, bh.SlotNumber)
-	enc = append(enc, buf...)
-	return enc
-}
-
-func (bh *BabeHeader) Decode(in []byte) error {
-	if len(in) < sr25519.VrfOutputLength+sr25519.VrfProofLength+16 {
-		return errors.New("input is too short: need at least VrfOutputLength (32) + VrfProofLength (64) + 16")
+func NewAuthorityData(pub *sr25519.PublicKey, weight uint64) *AuthorityData {
```
Comment (ryanchristo) [inline consensus/babe/types.go]: please add comment

File: consensus/babe/babe.go [hunk 3]
```diff
@@ -50,35 +51,51 @@ type Session struct {
 
 //nolint:structcheck
```
Comment (ryanchristo) [inline consensus/babe/babe.go]: Is this still needed?

File: consensus/babe/babe.go [hunk 4]
```diff
@@ -312,12 +356,12 @@ func (b *Session) buildBlockPreDigest(slot Slot) (*types.PreRuntimeDigest, error
 
 // buildBlockBabeHeader creates the BABE header for the slot.
 // the BABE header includes the proof of authorship right for this slot.
-func (b *Session) buildBlockBabeHeader(slot Slot) (*BabeHeader, error) {
+func (b *Session) buildBlockBabeHeader(slot Slot) (*babetypes.BabeHeader, error) {
```
Comment (ryanchristo) [inline consensus/babe/babe.go]: should this be `buildBabeBlockHeader`?

Comment (ryanchristo) [reply consensus/babe/babe.go]: oh wait, no I see now, vm

File: consensus/babe/babe.go (hunk 3 above)
Comment (noot) [reply consensus/babe/babe.go]: removed

File: consensus/babe/types.go (hunk 2 above)
Comment (noot) [reply consensus/babe/types.go]: done

File: cmd/gossamer/genesis.go [hunk 5]
```diff
@@ -77,6 +85,27 @@ func loadGenesis(ctx *cli.Context) error {
 	return t.Db().StoreGenesisData(gen)
 }
 
+// given raw genesis state data, return the initialized state trie and genesis block header.
+func initializeGenesisState(gen genesis.GenesisFields) (*trie.Trie, *types.Header, error) {
```
Comment (thomasmodeneis) [inline cmd/gossamer/genesis.go]: the comment should start always with the method name for the godocs to work properly, eg: // initializeGenesisState

File: consensus/babe/types/types.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+// BabeHeader as defined in Polkadot RE Spec, definition 5.10 in section 5.1.4
+type BabeHeader struct {
+	VrfOutput          [sr25519.VrfOutputLength]byte
+	VrfProof           [sr25519.VrfProofLength]byte
+	BlockProducerIndex uint64
+	SlotNumber         uint64
+}
+
+func (bh *BabeHeader) Encode() []byte {
```
Comment (thomasmodeneis) [inline consensus/babe/types/types.go]: new exported func should be have comments for godoc

File: consensus/babe/types/types.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	enc = append(enc, bh.VrfProof[:]...)
+	buf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(buf, bh.BlockProducerIndex)
+	enc = append(enc, buf...)
+	binary.LittleEndian.PutUint64(buf, bh.SlotNumber)
+	enc = append(enc, buf...)
+	return enc
+}
+
+func (bh *BabeHeader) Decode(in []byte) error {
```
Comment (thomasmodeneis) [inline consensus/babe/types/types.go]: new exported func should be have comments for godoc

File: consensus/babe/verify.go [hunk 8]
```diff
@@ -4,10 +4,11 @@ import (
 	"encoding/binary"
 	"fmt"
 
+	babetypes "github.com/ChainSafe/gossamer/consensus/babe/types"
 	"github.com/ChainSafe/gossamer/core/types"
 )
 
-func (b *Session) verifySlotWinner(slot uint64, header *BabeHeader) (bool, error) {
+func (b *Session) verifySlotWinner(slot uint64, header *babetypes.BabeHeader) (bool, error) {
```
Comment (thomasmodeneis) [inline consensus/babe/verify.go]: exported func should be have comments for godoc

File: state/block.go [hunk 9]
```diff
@@ -19,16 +21,45 @@ type blockState struct {
...
+	latestHeader, err := bs.GetHeader(latestHash)
+	if err != nil {
+		return bs, fmt.Errorf("NewBlockState latestBlock err: %s", err)
+	}
+
+	bs.latestHeader = latestHeader
+	return bs, nil
+}
+
+func NewBlockStateFromGenesis(dataDir string, header *types.Header) (*blockState, error) {
```
Comment (thomasmodeneis) [inline state/block.go]: exported func should be have comments for godoc

File: state/block.go [hunk 10]
```diff
@@ -219,3 +219,37 @@ func (bs *blockState) AddBlock(newBlock types.Block) error {
...
+func babeHeaderKey(epoch uint64, slot uint64) []byte {
+	epochBytes := make([]byte, 8)
+	binary.LittleEndian.PutUint64(epochBytes, epoch)
+	sliceBytes := make([]byte, 8)
+	binary.LittleEndian.PutUint64(sliceBytes, slot)
+	combined := append(epochBytes, sliceBytes...)
+	return append(babeHeaderPrefix, combined...)
+}
+
+func (bs *blockState) GetBabeHeader(epoch uint64, slot uint64) (*babetypes.BabeHeader, error) {
```
Comment (thomasmodeneis) [inline state/block.go]: exported func should be have comments for godoc

File: state/block.go [hunk 11]
```diff
@@ -219,3 +219,37 @@ func (bs *blockState) AddBlock(newBlock types.Block) error {
...
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(data, result)
+
+	return result, err
+}
+
+func (bs *blockState) SetBabeHeader(epoch uint64, slot uint64, blockData *babetypes.BabeHeader) error {
```
Comment (thomasmodeneis) [inline state/block.go]: exported func should be have comments for godoc

File: state/service.go [hunk 12]
```diff
@@ -20,6 +25,42 @@ func NewService(path string) *Service {
...
+
+	err = blockDb.db.Db.Close()
+	if err != nil {
+		return err
+	}
+
+	return storageDb.Db.Db.Close()
+}
+
 func (s *Service) Start() error {
```
Comment (thomasmodeneis) [inline state/service.go]: exported func should be have comments for godoc

File: state/storage.go [hunk 13]
```diff
@@ -12,14 +12,15 @@ type StorageState struct {
 	Db   *polkadb.StateDB
 }
 
-func NewStorageState(dataDir string) (*StorageState, error) {
+func NewStorageState(dataDir string, t *trie.Trie) (*StorageState, error) {
```
Comment (thomasmodeneis) [inline state/storage.go]: exported func should be have comments for godoc

File: state/storage.go [hunk 14]
```diff
@@ -63,6 +64,10 @@ func (s *StorageState) LoadFromDB(root common.Hash) error {
 	return s.trie.LoadFromDB(root)
 }
 
+func (s *StorageState) StoreInDB() error {
```
Comment (thomasmodeneis) [inline state/storage.go]: exported func should be have comments for godoc

File: trie/database.go [hunk 15]
```diff
@@ -56,12 +51,12 @@ func (db *Database) Load(key []byte) ([]byte, error) {
 	return db.Db.Get(key)
 }
 
-func (db *Database) StoreLatestHash(hash []byte) error {
-	return db.Db.Put(LatestHashKey, hash)
+func (db *Database) StoreLatestStorageHash(hash []byte) error {
```
Comment (thomasmodeneis) [inline trie/database.go]: exported func should be have comments for godoc

File: trie/database.go [hunk 16]
```diff
@@ -56,12 +51,12 @@ func (db *Database) Load(key []byte) ([]byte, error) {
...
+	return db.Db.Put(common.LatestStorageHashKey, hash)
 }
 
-func (db *Database) LoadLatestHash() (common.Hash, error) {
-	hashbytes, err := db.Db.Get(LatestHashKey)
+func (db *Database) LoadLatestStorageHash() (common.Hash, error) {
```
Comment (thomasmodeneis) [inline trie/database.go]: exported func should be have comments for godoc

File: trie/database.go [hunk 17]
```diff
@@ -75,11 +70,11 @@ func (db *Database) StoreGenesisData(gen *genesis.GenesisData) error {
 		return fmt.Errorf("cannot scale encode genesis data: %s", err)
 	}
 
-	return db.Store(GenesisDataKey, enc)
+	return db.Store(common.GenesisDataKey, enc)
 }
 
 func (db *Database) LoadGenesisData() (*genesis.GenesisData, error) {
```
Comment (thomasmodeneis) [inline trie/database.go]: exported func should be have comments for godoc

File: trie/trie.go [hunk 18]
```diff
@@ -60,6 +62,10 @@ func (t *Trie) Db() *Database {
 	return t.db
 }
 
+func (t *Trie) SetDb(db *Database) {
```
Comment (thomasmodeneis) [inline trie/trie.go]: exported func should be have comments for godoc

File: cmd/gossamer/genesis.go (hunk 5 above)
Comment (noot) [reply cmd/gossamer/genesis.go]: unexported functions don't get included in the godoc, so I'm not sure if it makes a difference. I can change it anyways

File: trie/trie.go (hunk 18 above)
Comment (noot) [reply trie/trie.go]: all comments have been added :)

**Pull request merged**